src/main/java/io/temporal/internal/sync/POJOWorkflowImplementationFactory.java

```diff
@@ -318,24 +296,76 @@ public void processSignal(String signalName, byte[] input, long eventId) {
-        newInstance();
-        signalMethod.invoke(workflow, args);
-      } catch (IllegalAccessException e) {
-        throw new Error("Failure processing \"" + signalName + "\" at eventID " + eventId, e);
+        Preconditions.checkNotNull(workflowInvoker, "initialize not called");
+        workflowInvoker.processSignal(signalName, args, eventId);
       } catch (DataConverterException e) {
         logSerializationException(signalName, eventId, e);
-      } catch (InvocationTargetException e) {
-        Throwable targetException = e.getTargetException();
-        if (targetException instanceof DataConverterException) {
-          logSerializationException(signalName, eventId, (DataConverterException) targetException);
-        } else if (targetException instanceof Error) {
-          throw (Error) targetException;
-        } else {
-          throw new Error(
-              "Failure processing \"" + signalName + "\" at eventID " + eventId, targetException);
+      }
+    }
+
+    private class RootWorkflowInvocationInterceptor implements WorkflowInvocationInterceptor {
+
+      @Override
+      public Object execute(Object[] arguments) {
+        WorkflowInfo context = Workflow.getWorkflowInfo();
+        newInstance();
+        WorkflowInternal.registerQuery(workflow);
+        try {
+          return workflowMethod.invoke(workflow, arguments);
+        } catch (IllegalAccessException e) {
+          throw new Error(mapToWorkflowExecutionException(e, dataConverter));
+        } catch (InvocationTargetException e) {
+          Throwable targetException = e.getTargetException();
+          if (targetException instanceof Error) {
+            throw (Error) targetException;
+          }
+          // Cancellation should be delivered as it impacts which decision closes a
+          // workflow.
+          if (targetException instanceof CancellationException) {
+            throw (CancellationException) targetException;
+          }
+          if (log.isErrorEnabled()) {
+            log.error(
+                "Workflow execution failure "
+                    + "WorkflowID="
+                    + context.getWorkflowId()
+                    + ", RunID="
+                    + context.getRunId()
+                    + ", WorkflowType="
+                    + context.getWorkflowType(),
+                targetException);
+          }
+          // Cast to Exception is safe as Error is handled above.
+          throw mapToWorkflowExecutionException((Exception) targetException, dataConverter);
+        }
+      }
+
+      @Override
+      public void init(WorkflowCallsInterceptor interceptor) {
+        WorkflowInternal.getRootDecisionContext().setHeadInterceptor(interceptor);
+      }
+
+      @Override
+      public void processSignal(String signalName, Object[] arguments, long eventId) {
+        Method signalMethod = signalHandlers.get(signalName);
+        try {
+          newInstance();
```

Comment:
I'm confused on why newInstance is created for both processSignal and execute calls.  Is it expected that only one of them will ever be called?

Reply:
It is a noop if the instance is already created. But you are right that after my latest changes it belongs in init call.